src/Analyzers/CSharp/Analysis/UseAsyncAwaitAnalyzer.cs

```diff
@@ -57,7 +57,7 @@ private static void AnalyzeMethodDeclaration(SyntaxNodeAnalysisContext context)
 
         IMethodSymbol methodSymbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration, context.CancellationToken);
 
-        if (!SymbolUtility.IsAwaitable(methodSymbol.ReturnType))
+        if (!methodSymbol.ReturnType.IsTaskType() || !methodSymbol.ReturnType.IsAwaitable(context.SemanticModel, body.SpanStart))
```

Comment:
Shouldn't the operator be `&&`?

Currently if `IsTaskType` returns true the second condition is checked which does not make sense (We already know it's awaitable since it's Task).

Reply:
Making a type "task-like" (i.e. applying the `AsyncMethodBuilder` attribute) doesn't actually require being awaitable, though I agree having a non-awaitable task-like type isn't very useful. Maybe there's some esoteric use for compilers like async state machines with custom behaviour, but that's way out of my depth.
```csharp
async TaskLike M()
{
    //await TaskLike; // no GetAwaiter
    await Task.CompletedTask;
}
```
For this analyzer in particular, we can't apply it unless the type is also awaitable.
```csharp
TaskLike M2() // can apply `async`, but...
{
    return TaskLike.Create(); // can't transform to `await ...`
}
```

Reply:
Oh, I see what you mean here.

I propose following change to ensure that if a type is well-known awaitable type we should bail out as early as possible:

Encapsulate this condition
```
methodSymbol.ReturnType.IsTaskLikeType()
    && methodSymbol.ReturnType.IsAwaitable(context.SemanticModel, context.Node.SpanStart))
```

into a method (name is just a suggestion):

```
SymbolUtility.HasAwaitableReturnType(IMethodSymbol ...
```

And ensure that condition

```
typeSymbol.ContainingNamespace.HasMetadataName(in MetadataNames.System_Threading_Tasks)
    && (typeSymbol.EqualsOrInheritsFrom(in MetadataNames.System_Threading_Tasks_Task)
        || typeSymbol.MetadataName is "ValueTask" or "ValueTask`1")
```

will be called either in that method or in `IsAwaitable`.